stwcs/wcsutil/altwcs.py

```diff
@@ -67,95 +81,362 @@ def archiveWCS(fname, ext, wcskey=" ", wcsname=" ", reusekey=False):
+        close_hdulist = True
     else:
-        f = fname
+        h = fname
+        close_hdulist = False
 
-    if not _parpasscheck(f, ext, wcskey, wcsname):
-        closefobj(fname, f)
-        raise ValueError("Input parameters problem")
+        # validate file object and open mode:
+        if not isinstance(h, fits.HDUList):
+            if close_hdulist:
+                h.close()
+            raise TypeError("'fname' must be either a FITS file object or a string file name.")
 
-    # Interpret input 'ext' value to get list of extensions to process
-    ext = _buildExtlist(f, ext)
+        if h.fileinfo(0) is not None and h.fileinfo(0)['filemode'] != 'update':
+            if close_hdulist:
+                h.close()
+            raise ValueError("'fname' must be a file object opened in 'update' mode.")
+
+    # validate and interpret extension(s):
+    try:
+        if isinstance(ext, list):
+            map(_validate_ext, ext)
+        else:
+            _validate_ext(ext)
+            ext = [ext]
 
-    if wcsname == ' ':
-        wcsname = wcs_from_key(f, ext[0]).get('WCSNAME', ' ')
+    except ValueError as e:
+        if close_hdulist:
+            h.close()
+        raise e
 
     wcsext = ext[0]
-    if wcskey != " " and wcskey in wcskeys(f[wcsext].header) and not reusekey:
-        closefobj(fname, f)
-        raise KeyError(f"Wcskey '{wcskey}' is aready used. \
-        Run archiveWCS() with reusekey=True to overwrite this alternate WCS. \
-        Alternatively choose another wcskey with altwcs.available_wcskeys().")
-
-    elif wcskey == " ":
-        # wcsname exists, overwrite it if reuse is True or get the next key
-        if wcsname in _alt_wcs_names(f[wcsext].header).values():
-            if reusekey:
-                # try getting the key from an existing WCS with WCSNAME
-                wkey = getKeyFromName(f[wcsext].header, wcsname)
-                wname = wcsname
-                if wkey == ' ':
-                    wkey = _next_wcskey(f[wcsext].header)
-                elif wkey is None:
-                    closefobj(fname, f)
-                    raise KeyError(f"Could not get a valid wcskey from wcsname '{wcsname:s}'")
+    hdr = h[wcsext].header
+
+    wcs_names = _alt_wcs_names(hdr)
+    wcs_names_u = {k: v.upper() for k, v in wcs_names.items()}
+    wcsname_u = None if wcsname is None else wcsname.upper()
+    wcs_keys = wcskeys(h, wcsext)
+
+    # validate and process mode:
+    if mode & ~sum(AM):
+        raise ValueError("Unrecognized 'mode' flag.")
+    overwrite_key = bool(mode & AM.OVERWRITE_KEY)
+    auto_rename = bool(mode & AM.AUTO_RENAME)
+
+    wcs_id = (None, None)
+
+    # do not overwrite OPUS WCS if present regardless of 'mode':
+    if ((wcskey == 'O' and wcskey in wcs_keys) or
+            (wcsname == 'OPUS' and wcsname in wcs_names.values())):
+        if close_hdulist:
+            h.close()
+        log.debug("WCS key 'O' already exists and cannot be overwritten.")
```

Comment:
If I am not mistaken (and I may be), this logic would report that the OPUS altWCS already exists even if the OPUS altWCS had a key that was not 'O'.  That is fine, but it may confuse the user if they get this message under that situation.

Reply:
Good catch! The message is confusing (i.e., unrelated to wcsname being OPUS). However, are you OK with not allowing users to archive a WCS under `'OPUS'` name? The reason I wanted to do this because I no longer want to allow duplicates in Alt WCS and so the issue is: is it OK for us to "reserve" not only the `'O'` _key_ but also the `'OPUS'` _wcsname_?

If that is OK, I will go ahead and clarify the error message.

Reply:
We should reserve both the key 'O' and the corresponding wcsname 'OPUS' as special WCS, since we need to preserve this information at all costs in order for the updatewcs code to work at all.  By reserving both, we guarantee that this information will at least have an altWCS to use for this information at all times while making it easy for us to document for the end user.